File: chroma-league/src/test/java/com/bonepl/chromaleague/hud/animations/AnimationTester.java
package com.bonepl.chromaleague.hud.animations;


import com.bonepl.razersdk.ChromaRestSDK;
import com.bonepl.razersdk.animation.IFrame;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.function.Consumer;
import java.util.function.Supplier;

public final class AnimationTester {
    private static final Logger LOGGER = LogManager.getLogger();
    private Consumer<Integer> afterIterationAction;
    private int sleepTime = 50;
    private int sleepAfter;

    public AnimationTester withAfterIterationAction(Consumer<Integer> action) {
        afterIterationAction = action;
        return this;
    }

    public AnimationTester withSleepTime(int milliseconds) {
        sleepTime = milliseconds;
        return this;
    }

    public AnimationTester withSleepAfter(int milliseconds) {
        sleepAfter = milliseconds;
        return this;
    }

    public void testAnimation(Supplier<? extends IFrame> supplier, int iterations) {
        try (ChromaRestSDK chromaRestSDK = new ChromaRestSDK()) {
            for (int i = 0; i < iterations; i++) {
                final IFrame frame = supplier.get();
                chromaRestSDK.createKeyboardEffect(frame);
                executeAfterIteration(i);
                sleepThread(sleepTime);
            }
            sleepThread(sleepAfter);
        }
    }

    public void testAnimation(IFrame frame) {
        try (ChromaRestSDK chromaRestSDK = new ChromaRestSDK()) {
            for (int i = 0; frame.hasFrame(); i++) {
                chromaRestSDK.createKeyboardEffect(frame);
                executeAfterIteration(i);
                sleepThread(sleepTime);
            }
            sleepThread(sleepAfter);
        }
    }

    public void testAnimation(IFrame frame, int iterations) {
        try (ChromaRestSDK chromaRestSDK = new ChromaRestSDK()) {
            for (int i = 0; i < iterations; i++) {
                chromaRestSDK.createKeyboardEffect(frame);
                executeAfterIteration(i);
                sleepThread(sleepTime);
            }
            sleepThread(sleepAfter);
        }
    }

    private void executeAfterIteration(int i) {
        if (afterIterationAction != null) {
            afterIterationAction.accept(i);
        }
    }

    private static void sleepThread(int milliseconds) {
        try {
            Thread.sleep(milliseconds);
        } catch (InterruptedException e) {
            LOGGER.error(e);
        }
    }
}

File: test/AnimationTester.ts
import {ChromaNativeSDK} from '../src/razer-sdk/ChromaNativeSDK.js';
import {IFrame} from '../src/razer-sdk/animation/IFrame.js';

/**
 * Helper class for testing animations on actual hardware.
 * Uses native DLL for reliable communication with Razer Chroma SDK.
 */
export class AnimationTester {
    private beforeIterationAction?: (i: number) => void;
    private afterIterationAction?: (i: number) => void;
    private sleepTime: number = 50;
    private sleepAfter: number = 0;

    withBeforeIterationAction(action: (i: number) => void): this {
        this.beforeIterationAction = action;
        return this;
    }

    withAfterIterationAction(action: (i: number) => void): this {
        this.afterIterationAction = action;
        return this;
    }

    withSleepTime(milliseconds: number): this {
        this.sleepTime = milliseconds;
        return this;
    }

    withSleepAfter(milliseconds: number): this {
        this.sleepAfter = milliseconds;
        return this;
    }

    /**
     * Test animation until hasFrame returns false
     */
    async testAnimation(frame: IFrame, iterations?: number): Promise<void> {
        const sdk = new ChromaNativeSDK();
        await sdk.connect();

        try {
            if (iterations !== undefined) {
                for (let i = 0; i < iterations; i++) {
                    this.beforeIterationAction?.(i);
                    await sdk.createKeyboardEffect(frame);
                    this.afterIterationAction?.(i);
                    await this.sleep(this.sleepTime);
                }
            } else {
                let i = 0;
                while (frame.hasFrame()) {
                    this.beforeIterationAction?.(i);
                    await sdk.createKeyboardEffect(frame);
                    this.afterIterationAction?.(i);
                    await this.sleep(this.sleepTime);
                    i++;
                }
            }

            await this.sleep(this.sleepAfter);
        } finally {
            await sdk.close();
        }
    }

    private sleep(milliseconds: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, milliseconds));
    }
}
